parsl/providers/local/local.py

```diff
@@ -76,19 +76,70 @@ def status(self, job_ids):
+            # This job dict should really be a class on its own
+            job_dict = self.resources[job_id]
+            if job_dict['status'] is not None and job_dict['status'].terminal:
+                # We already checked this and it can't change after that
+                continue
+            # Script path should point to remote path if _should_move_files() is True
+            script_path = job_dict['script_path']
+
+            alive = self._is_alive(job_dict)
+            str_ec = self._read_job_file(script_path, '.ec').strip()
+
+            status = None
+            if str_ec == '-':
+                if alive:
+                    status = JobStatus(JobState.RUNNING)
+                else:
+                    # not alive but didn't get to write an exit code
+                    if 'cancelled' in job_dict:
+                        # because we cancelled it
+                        status = JobStatus(JobState.CANCELLED)
+                    else:
+                        # we didn't cancel it, so it must have been killed by something outside
+                        # parsl; we don't have a state for this, but we'll use CANCELLED with
+                        # a specific message
+                        status = JobStatus(JobState.CANCELLED, message='Killed')
+            else:
+                try:
+                    # TODO: ensure that these files are only read once and clean them
+                    ec = int(str_ec)
+                    out = self._read_job_file(script_path, '.out')
+                    err = self._read_job_file(script_path, '.err')
+                    if ec == 0:
+                        state = JobState.COMPLETED
                     else:
-                        self.resources[job_id]['status'] = JobStatus(JobState.FAILED)
+                        state = JobState.FAILED
+                    status = JobStatus(state, exit_code=ec, stdout=out, stderr=err)
+                except Exception:
+                    status = JobStatus(JobState.FAILED,
+                                       'Cannot parse exit code: {}'.format(str_ec))
+
+            job_dict['status'] = status
 
         return [self.resources[jid]['status'] for jid in job_ids]
 
+    def _is_alive(self, job_dict):
+        retcode, stdout, stderr = self.channel.execute_wait(
+            'ps -p {} > /dev/null 2> /dev/null; echo "STATUS:$?" '.format(
+                job_dict['remote_pid']), self.cmd_timeout)
+        for line in stdout.split('\n'):
+            if line.startswith("STATUS:"):
+                status = line.split("STATUS:")[1].strip()
+                if status == "0":
+                    return True
+                else:
+                    return False
+
+    def _read_job_file(self, script_path: str, suffix: str) -> str:
+        path = '{0}{1}'.format(script_path, suffix)
+        if self._should_move_files():
+            path = self.channel.pull_file(path, self.script_dir)
+
+        with open(path, 'r') as f:
```

Comment:
It looks like this will only be run on status files which would be small. Can you confirm please?

Reply:
No such guarantee. We can possibly change to a lazy read for stdout/stderr.

Reply:
Lazy reading is now being done

Reply:
After some more discussion, I added stdout_summary and stderr_summary properties in JobStatus. They read first and last 1k of the respective files.

By the way, the stdout/stderr properties are only used by test_local_provider.py, so the _summary methods are not really needed anywhere.